packages/snowflake/src/lib/Snowflake.ts

```diff
@@ -173,7 +173,7 @@ export class Snowflake {
 	 * @returns The UNIX timestamp that is stored in `id`.
 	 */
 	public timestampFrom(id: string | bigint): number {
-		return Number((BigInt(id) >> 22n) + this[EpochSymbol]);
+		return Math.floor(Number(id) / 2 ** 22) + Number(this[EpochSymbol]);
```

Comment:
Two suggestions:

- Regarding `2 ** 22`:
  - What’s the performance if we replace it with `1 << 22`?
  - What if we made it a constant?
- Regarding `Number(this[EpochSymbol])`:
  - What if we make `EpochNumberSymbol` which is the result as a number?

Reply:
Fastest seems to be caching the power of 2 multiplication, and having numeric epochs

```
┌─────────┬────────────────────────────────────────────────────┬──────────────┬────────────────────┬──────────┬─────────┐
│ (index) │ Task Name                                          │ ops/sec      │ Average Time (ns)  │ Margin   │ Samples │
├─────────┼────────────────────────────────────────────────────┼──────────────┼────────────────────┼──────────┼─────────┤
│ 0       │ 'bigint shift + numeric conversion'                │ '9,620,374'  │ 103.94605202702998 │ '±1.80%' │ 962038  │
│ 1       │ 'division with cached power of 2 and epoch'        │ '26,279,831' │ 38.0519946088112   │ '±0.58%' │ 2627984 │
│ 2       │ 'division with cached left shift and epoch'        │ '25,828,721' │ 38.716589240114665 │ '±0.59%' │ 2582873 │
│ 3       │ 'division with power of 2 and epoch'               │ '25,573,364' │ 39.1031850708978   │ '±0.67%' │ 2557337 │
│ 4       │ 'division with left shift and epoch'               │ '25,703,634' │ 38.90500372692088  │ '±0.87%' │ 2570364 │
│ 5       │ 'division with cached power of 2 and bigint epoch' │ '16,038,767' │ 62.34892887667072  │ '±6.64%' │ 1603877 │
│ 6       │ 'numeric left shift and epoch'                     │ '25,424,629' │ 39.33193980271842  │ '±1.07%' │ 2542464 │
│ 7       │ 'numeric left shift and bigint epoch'              │ '21,102,698' │ 47.3873035200918   │ '±0.58%' │ 2110270 │
└─────────┴────────────────────────────────────────────────────┴──────────────┴────────────────────┴──────────┴─────────┘
```

<details>
<summary>Test Code</summary>

```ts
import { Bench } from 'tinybench';

const bench = new Bench({ time: 100 });

const id = '139836912335716352';
const discordEpoch = 1_420_070_400_000n;
const numericDiscordEpoch = 1_420_070_400_000;
const powerOfTwoCached = 2 ** 22;
const leftShiftCached = 1 << 22;

bench.add('bigint shift + numeric conversion', () => {
	Number((BigInt(id) >> 22n) + discordEpoch);
});

bench.add('division with cached power of 2 and epoch', () => {
	Math.floor(Number(Number(id) / powerOfTwoCached + numericDiscordEpoch));
});

bench.add('division with cached left shift and epoch', () => {
	Math.floor(Number(Number(id) / leftShiftCached + numericDiscordEpoch));
});

bench.add('division with power of 2 and epoch', () => {
	Math.floor(Number(Number(id) / 2 ** 22 + numericDiscordEpoch));
});

bench.add('division with left shift and epoch', () => {
	Math.floor(Number(Number(id) / (1 << 22) + numericDiscordEpoch));
});

bench.add('division with cached power of 2 and bigint epoch', () => {
	Math.floor(Number(Number(id) / powerOfTwoCached + Number(discordEpoch)));
});

bench.add('numeric left shift and epoch', () => {
	Math.floor(Number((Number(id) >> 22) + numericDiscordEpoch));
});

bench.add('numeric left shift and bigint epoch', () => {
	Math.floor(Number((Number(id) >> 22) + Number(discordEpoch)));
});

await bench.warmup();
await bench.run();

console.table(bench.table());
```

</details>

Reply:
One thing to keep in mind: numeric left shfits *lose* precision! They cannot be used, I just threw them in for completeness sake